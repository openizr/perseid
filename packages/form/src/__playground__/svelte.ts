/* c8 ignore start */
/* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call */

import { SvelteComponent } from 'svelte';
import Engine from 'scripts/core/Engine';
import Form from 'scripts/svelte/Form.svelte';
import Field from '__playground__/Field.svelte';

let app: Form;
const { log } = console;

function main(): void {
  app = new Form({
    props: {
      Field: Field as typeof SvelteComponent | null | undefined,
      configuration: {
        root: 'root',
        fields: {
          test: { type: 'string', required: true },
        },
        steps: {
          root: {
            fields: ['test'],
          },
        },
        plugins: [
          (engine: Engine): void => {
            engine.on('userAction', (data, next) => {
              log(data);
              return next(data);
            });
          },
        ],
      },
    },
    target: document.getElementById('root') as unknown as HTMLElement,
  });
}

// Ensures DOM is fully loaded before running app's main logic.
// Loading hasn't finished yet...
if (document.readyState === 'loading') {
  document.addEventListener('DOMContentLoaded', main);
  // `DOMContentLoaded` has already fired...
} else {
  main();
}

// Ensures subscriptions to Store are correctly cleared when page is left, to prevent "ghost"
// processing, by manually unmounting Vue components tree.
window.addEventListener('beforeunload', () => {
  (app as { $destroy: () => void; }).$destroy();
});
