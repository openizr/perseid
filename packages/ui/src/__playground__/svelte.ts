/* c8 ignore start */

import type { SvelteComponent } from 'svelte';
import Router from '__playground__/pages/Router.svelte';

let app: SvelteComponent;

function main(): void {
  const target = document.querySelector('#root') as unknown as HTMLElement;
  target.innerHTML = '';
  app = new Router({
    hydrate: false,
    target,
  });
}

// Ensures DOM is fully loaded before running app's main logic.
// Loading hasn't finished yet...
if (document.readyState === 'loading') {
  document.addEventListener('DOMContentLoaded', main);
  // `DOMContentLoaded` has already fired...
} else {
  main();
}

// Ensures subscriptions to Store are correctly cleared when page is left, to prevent "ghost"
// processing, by manually unmounting Svelte components tree.
window.addEventListener('beforeunload', () => {
  app.$destroy();
});
