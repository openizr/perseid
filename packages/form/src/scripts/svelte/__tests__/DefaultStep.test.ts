/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @vitest-environment jsdom
 */

import type Engine from 'scripts/core/Engine';
import { type SvelteComponent } from 'svelte';
import { render } from '@testing-library/svelte';
import DefaultStep from 'scripts/svelte/DefaultStep.svelte';
import DefaultField from 'scripts/svelte/DefaultField.svelte';

describe('svelte/DefaultStep', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('renders correctly', () => {
    const step: Step = {
      path: 'root.0',
      status: 'initial',
      fields: [{
        path: 'root.0.test',
        status: 'initial',
        error: null,
        value: null,
        required: false,
        type: 'string',
      }],
    };
    const { container } = render(DefaultStep, {
      props: {
        step,
        active: false,
        useSubscription: vi.fn(),
        engine: {} as unknown as Engine,
        Field: DefaultField as unknown as typeof SvelteComponent,
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
