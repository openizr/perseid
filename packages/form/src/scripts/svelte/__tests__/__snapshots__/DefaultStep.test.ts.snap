// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`svelte/DefaultStep > renders correctly 1`] = `
<div>
  <div
    class="perseid-form__step__fields"
  >
    
    <pre>
        
      {
  "path": "root.0.test",
  "type": "string",
  "status": "initial",
  "error": null,
  "active": false,
  "required": false,
  "value": null,
  "Field": "&lt;Component&gt;",
  "engine": "&lt;Engine&gt;",
  "useSubscription": "&lt;Function&gt;"
}
      

    </pre>
    <!--&lt;DefaultField&gt;-->
    
    
  </div>
  <!--&lt;DefaultStep&gt;-->
</div>
`;
